### ISSUE TITLE
destination syslog forget messages on long disconnect/reconnect

### ISSUE THREAD

**Issue by asmodehn** [syslog-ng/syslog-ng#5021]
# syslog-ng
## Version of syslog-ng
```
syslog-ng 4 (4.7.1)
Config version: 4.2
Installer-Version: 4.7.1
Revision:
Compile-Date: Jun 19 2024 08:24:49
Module-Directory: /usr/lib/syslog-ng
Module-Path: /usr/lib/syslog-ng
Include-Path: /usr/share/syslog-ng/include
Available-Modules: add-contextual-data,affile,afprog,afsocket,afstomp,afuser,appmodel,azure-auth-header,basicfuncs,cef,confgen,correlation,cryptofuncs,csvparser,disk-buffer,examples,graphite,hook-commands,json-plugin,kvformat,linux-kmsg-format,map-value-pairs,metrics-probe,pseudofile,rate-limit-filter,regexp-parser,secure-logging,stardate,syslogformat,system-source,tags-parser,tfgetent,timestamp,xml
Enable-Debug: off
Enable-GProf: off
Enable-Memtrace: off
Enable-IPv6: on
Enable-Spoof-Source: off
Enable-TCP-Wrapper: off
Enable-Linux-Caps: off
Enable-Systemd: off
```

## Platform
`Linux myhostname 6.6.33 #0 SMP Wed Jun 19 08:24:49 2024 aarch64 GNU/Linux`
This is a Busybox (OpenWRT)
 

# Issue
## Failure

syslog-ng is connected to a remote syslog destination and sending messages.
The networks breaks down (I bring network interface down to trigger it).
```
# logread -f
Jul 16 07:54:10 myhostname wpa_supplicant[9807]: Remove interface phy0-sta0
Jul 16 07:54:11 myhostname wpa_supplicant[9807]: phy0-sta0: CTRL-EVENT-DISCONNECTED bssid=04:42:1a:6c:72:a8 reason=3 locally_generated=1
Jul 16 07:54:11 myhostname wpa_supplicant[9807]: phy0-sta0: CTRL-EVENT-DSCP-POLICY clear_all
Jul 16 07:54:11 myhostname wpa_supplicant[9807]: phy0-sta0: CTRL-EVENT-DSCP-POLICY clear_all
Jul 16 07:54:11 myhostname wpa_supplicant[9807]: nl80211: deinit ifname=phy0-sta0 disabled_11b_rates=0
Jul 16 07:54:11 myhostname NET: hotplug (iface): action='remove' interface='phy0-sta0'
Jul 16 07:54:11 myhostname ModemManager[27637]: hotplug: remove network interface phy0-sta0: event processed
```
syslog-ng fills up the TCP SendQ. Until the socket dies(?).
```
# watch netstat -tap | grep syslog-ng
tcp     3542  62919 192.168.50.95:24680     my.server.com:10044      ESTABLISHED 22625/syslog-ng
tcp     3542  63625 192.168.50.95:24680     my.server.com:10044      ESTABLISHED 22625/syslog-ng
tcp     3542  63625 192.168.50.95:24680     my.server.com:10044      ESTABLISHED 22625/syslog-ng
tcp     3542  64331 192.168.50.95:24680     my.server.com:10044      ESTABLISHED 22625/syslog-ng
```
syslog-ng finally detects disconnection.
```
# logread -f
Jul 16 08:10:36 myhostname syslog-ng[22625]: Syslog connection broken; fd='19', server='AF_INET(my.server.com:10044)', time_reopen='60'
```
syslog-ng starts to queue messages.
```
# watch "( syslog-ng-ctl stats | grep 10044 )"
dst.syslog;logstash_syslog_tcp#0;tls,my.server.com:10044;a;dropped;0
dst.syslog;logstash_syslog_tcp#0;tls,my.server.com:10044;a;queued;18
dst.syslog;logstash_syslog_tcp#0;tls,my.server.com:10044;a;written;11992
dst.syslog;logstash_syslog_tcp#0;tls,my.server.com:10044;a;memory_usage;12204
dst.syslog;logstash_syslog_tcp#0;tls,my.server.com:10044;a;msg_size_avg;305
```

But this is too late : The messages on the TCP SendQ are lost.
They are not in syslog-ng queue and will not be sent again when connection is recovered.

## Additional remarks
- The message rate is quite slow (2 every 10 seconds or so), so throughput is not an issue.
- I had the problem with a "non-reliable" diskbuffer before, I simplified the configuration to reproduce only with a usual memory buffer.
- The flow-control is disabled, but, as far as I understand it, the problem comes before the need for flowcontrol to notify the source... It occurs when syslog-ng sees (too late) that it needs to queue messages for the destination...

## Steps to reproduce
- Configure syslog-ng to slowly send messages to a remote syslog destination.
- Interrupt the network, while watching the TCP Send-Q
- Notice syslog-ng actually only see the disconnection when the socket queue is full.
- Notice the messages are not queued by syslog-ng until disconnection is detected.
- Recover network, and notice the first messages sent during the early disconnection phase (while not detected by syslog-ng) were never sent.

## Configuration
```
# cat /etc/syslog-ng.conf
# Collect all local logs into a single file /var/log/messages.
# See https://www.syslog-ng.com/technical-documents/list/syslog-ng-open-source-edition

@version: 4.7
@include "scl.conf"

options {
        chain_hostnames(no); # Enable or disable the chained hostname format.
        create_dirs(yes);
        keep_hostname(yes); # Enable or disable hostname rewriting.
        log_fifo_size(256); # The number of messages that the output queue can store.
        log_msg_size(1024); # Maximum length of a message in bytes.
        stats(freq(0)); # The period between two STATS messages (sent by syslog-ng, containing statistics about dropped logs) in
        flush_lines(0); # How many lines are flushed to a destination at a time.
        use_fqdn(no); # Add Fully Qualified Domain Name instead of short hostname.
        use_dns(no);
};

# syslog-ng gets messages from syslog-ng (internal) and from /dev/log

source src {
        internal();
        unix-dgram("/dev/log");
};

source net {
        network_localhost();
};

source s_network {
        default-network-drivers(
                # NOTE: TLS support
                #
                # the default-network-drivers() source driver opens the TLS
                # enabled ports as well, however without an actual key/cert
                # pair they will not operate and syslog-ng would display a
                # warning at startup.
                #
                #tls(key-file("/path/to/ssl-private-key") cert-file("/path/to/ssl-cert"))
        );
};

source kernel {
        file("/proc/kmsg" program_override("kernel"));
};

destination messages {
        file(
                # Name cannot change, for logread to work as expected.
                "/var/log/messages"
        );
};


source namedpipe {
        pipe("/dev/my_namedpipe");
};

destination logstash_syslog_tcp {
        syslog("my.server.com"
                transport("tls")
                port(10044)
                tls(
                        key-file("/etc/x509/logstash-oss.key.pem")
                        cert-file("/etc/x509/logstash-oss.cert.pem")
                        ca-dir("/etc/x509")
                        peer-verify(no)
                )
                # currently we want it "CSV-like"
                # beginning of message (from syslog) in first cell
                template(",${HOST},${MESSAGE}")

                # Memory buffering
                # size in nb of messages
                log-fifo-size(1000)
        );
};


log {
        source(namedpipe);

        # TMP: for debugging only
        destination {
                file("/var/log/piped_messages");
        };

        # uncomment this to send logs remotely
        destination(logstash_syslog_tcp);

        # no flow-control. if buffer is too small,
        # we prefer syslog dropping messages
        # rather than potentially blowing up our source
        # flags(flow-control);

        flags(final);
};


log {
        source(src);
        source(net);
        # do not log kernel messages. use dmesg instead.
        #source(kernel);
        destination(messages);

        # uncomment this line to open port 514 to receive messages
        #source(s_network);
};
```

**Comment by HofiOne** [comment]
Hi,
Thank you for reporting this!
Could you please replace the Configuration with the simplest possible one that reproduces the concrete issue, as I can see it contains a lot of not-really related entries, I will try to reproduce the issue based on your Steps to reproduce instructions, but it would be nice to see the config you've used when discovered the problem.
TIA!

**Comment by HofiOne** [comment]
Just for the record

A simple config that can reproduce the problem

```

@version: 4.7
@include "scl.conf"

source src {
        file("/tmp/input.log");
};

destination dest {
        syslog("target_host"
                port(514)                
        );
};


log {
        source(src);

        destination(dest);
};


```

**Comment by HofiOne** [comment]
I guess this is how the current syslog destination, just like any other socket-based destinations works (even the TCP-based ones)

As you stated, `syslog-ng fills up the TCP SendQ` for a while if an already connected socket gets broken meanwhile communication goes through it.

This is happening in a layer that syslog-ng knows about nothing, it is signaled to it that its socket writing was successful for a while (even if the connection itself might have already been lost meanwhile), and the underlying layers buffer the written data (depending on the system configurations and the size of the data is sent, the buffering can accept a huge amount of data), there is no dedicated application layer acknowledgment mechanism that can notify syslog-ng about the loss of data that is stored in buffers of the lower layers.

You can try to decrease the detection time of the connection loss setting the various [so-keepalive(), tcp-keepalive-intvl(), tcp-keepalive-probes(), tcp-keepalive-time()](https://syslog-ng.github.io/admin-guide/070_Destinations/300_syslog/000_syslog_destination_options) options, but those are not working on all systems and still will not guarantee a data loss like this I think.

OTLP might be your best bet for a more reliable solution as it is described e.g. [here](https://axoflow.com/ingesting-large-volume-data-stream-to-the-cloud/).

**Comment by asmodehn** [comment]
Thanks a lot for checking.

After managing to reproduce the problem between two debian machines, with a recent syslog-ng built from source, I reached a similar conclusion.

Interestingly, I noticed the error was detected earlier and queue didn't fill up as much on my debian setup... although the tcp_keepalive settings are the same as on my OpenWRT setup... so something else (independent from syslog-ng) is amiss on my custom OpenWRT setup.

In any case, since I expect all messages to eventually arrive, I should indeed use a protocol that provide ACKs at the application level.

Thanks for the help !
